### PR TITLE
#19 Allow to break build on preview analysis runs

### PR DESCRIPTION
**Pull request by ghenzler**

### CODE REVIEW

**Comment by crgardner** [comment]
Anything else need to be done to get this to fail a gradle build?  I have a project with a quality gate and am running the following:

* SonarQube 6.1
* gradle with org.sonarqube plugin in preview mode
* Build Breaker plugin configured

I purposefully introduced violations to try to break build.  HTML issues report shows two problems.  However, Gradle build doesn't fail.  I noticed this is the debug log:

"QualityGateBreaker is disabled (sonar.analysis.mode != publish)"

Any thoughts on how to fix?

**Comment by ghenzler** [comment]
See comment  https://github.com/SonarQubeCommunity/sonar-build-breaker/issues/19#issuecomment-264049384 - the name "buildbreaker" is hardcoded to be excluded by default. The message "QualityGateBreaker is disabled (sonar.analysis.mode != publish)" comes from https://github.com/SonarQubeCommunity/sonar-build-breaker/blob/master/src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java#L85 (which is fine and stays like that), this PR introduces an entirely new PostJob (https://github.com/ghenzler/sonar-build-breaker/blob/feature/%2319/src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java#L35)

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  private final int failForIssueSeverityAsInt;
+  
+  String failureMessage = null;
+  
+  public BasicIssuesBuildBreaker(AnalysisMode analysisMode, ProjectIssues projectIssues, Settings settings) {
+    this.analysisMode = analysisMode;
+    this.projectIssues = projectIssues;
+    
+    failForIssuesWithSeverity = settings.getString(BuildBreakerPlugin.FAIL_FOR_ISSUES_WITH_SEVERITY_KEY);
+    failForIssueSeverityAsInt = Severity.ALL.indexOf(failForIssuesWithSeverity.trim().toUpperCase());
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: I don't think trim() is necessary

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  private static final String CLASSNAME = BasicIssuesBuildBreaker.class.getSimpleName();
+
+  private static final Logger LOG = Loggers.get(ForbiddenConfigurationBreaker.class);
+
+  private final ProjectIssues projectIssues;
+  private final AnalysisMode analysisMode;
+  private final String failForIssuesWithSeverity;
+  private final int failForIssueSeverityAsInt;
+  
+  String failureMessage = null;
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Should be private

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java (hunk 1 above)
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Locale should be specified when using toUpperCase()

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    
+    failForIssuesWithSeverity = settings.getString(BuildBreakerPlugin.FAIL_FOR_ISSUES_WITH_SEVERITY_KEY);
+    failForIssueSeverityAsInt = Severity.ALL.indexOf(failForIssuesWithSeverity.trim().toUpperCase());
+    
+  }
+
+  @Override
+  public void executeOn(Project project, SensorContext context) {
+    if (analysisMode.isPublish()) {
+      LOG.debug("{} is disabled ({} == {})", CLASSNAME, CoreProperties.ANALYSIS_MODE, CoreProperties.ANALYSIS_MODE_PUBLISH);
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Long line

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      return;
+    }
+
+    int issueCountToFailFor = 0;
+    for(Issue issue: projectIssues.issues()) {
+      String severity = issue.severity();
+      int issueSeverityAsInt = Severity.ALL.indexOf(String.valueOf(severity).trim().toUpperCase());
+      if (issueSeverityAsInt >= failForIssueSeverityAsInt) {
+        LOG.info("Recording issue {} that has a severity of '{}'",
+            issue.key(), severity);
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Combine statement's lines

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    
+    if(failForIssueSeverityAsInt < 0) {
+      LOG.debug("failForIssuesWithSeverity is configured to {}", failForIssuesWithSeverity);
+      return;
+    }
+
+    int issueCountToFailFor = 0;
+    for(Issue issue: projectIssues.issues()) {
+      String severity = issue.severity();
+      int issueSeverityAsInt = Severity.ALL.indexOf(String.valueOf(severity).trim().toUpperCase());
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Same notes on trim and toUpperCase

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      String msg = "Project " + project.getName() + " has " + issueCountToFailFor
+          + " issues that are of severity equal or higher than " + failForIssuesWithSeverity;
+      LOG.debug(msg);
+      // only mark failure and fail on PostJobsPhaseHandler.onPostJobsPhase() to ensure other plugins can finish their work, most notably the stash issue reporter plugin
+      markFailure(msg);
+      
+    } else {
+      LOG.info("Project " + project.getName() + " has no issues with severity equal or higher than " + failForIssuesWithSeverity);
+    }
+
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Remove extra line

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  @Override
+  public void onPostJobsPhase(PostJobsPhaseEvent event) {
+    if(event.isEnd() && failureMessage!=null) {
+      LOG.error(BuildBreakerPlugin.BUILD_BREAKER_LOG_STAMP+" "+failureMessage);
+      throw new IllegalStateException(failureMessage);
+    }
+  }
+
+}
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: End file with a newline

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  @Override
+  public void onPostJobsPhase(PostJobsPhaseEvent event) {
+    if(event.isEnd() && failureMessage!=null) {
+      LOG.error(BuildBreakerPlugin.BUILD_BREAKER_LOG_STAMP+" "+failureMessage);
+      throw new IllegalStateException(failureMessage);
+    }
+  }
+
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Remove extra line

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  }
+  
+  public void markFailure(String failureMessage) {
+    this.failureMessage = failureMessage;
+  }
+
+  @Override
+  public void onPostJobsPhase(PostJobsPhaseEvent event) {
+    if(event.isEnd() && failureMessage!=null) {
+      LOG.error(BuildBreakerPlugin.BUILD_BREAKER_LOG_STAMP+" "+failureMessage);
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Spaces around "+" tokens

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    
+  }
+  
+  public void markFailure(String failureMessage) {
+    this.failureMessage = failureMessage;
+  }
+
+  @Override
+  public void onPostJobsPhase(PostJobsPhaseEvent event) {
+    if(event.isEnd() && failureMessage!=null) {
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Space after "if" token

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+      // only mark failure and fail on PostJobsPhaseHandler.onPostJobsPhase() to ensure other plugins can finish their work, most notably the stash issue reporter plugin
+      markFailure(msg);
+      
+    } else {
+      LOG.info("Project " + project.getName() + " has no issues with severity equal or higher than " + failForIssuesWithSeverity);
+    }
+    
+  }
+  
+  public void markFailure(String failureMessage) {
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: private

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java [hunk 12]
```diff
@@ -83,8 +93,11 @@
 
   public static final String ALTERNATIVE_SERVER_URL_KEY = "sonar.buildbreaker.alternativeServerUrl";
 
+  public static final String FAIL_FOR_ISSUES_WITH_SEVERITY_KEY = "sonar.buildbreaker.preview.failForIssuesWithSeverity";
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: Too verbose.  `failAtSeverity`

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java [hunk 13]
```diff
@@ -83,8 +93,11 @@
...
+  public static final String FAIL_FOR_ISSUES_DISABLED = "DISABLED";
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: DISABLED would be a good enough name.  But is it not possible to use a blank string as the disabled option instead?

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+      LOG.debug("failForIssuesWithSeverity is configured to {}", failForIssuesWithSeverity);
+      return;
+    }
+
+    int issueCountToFailFor = 0;
+    for(Issue issue: projectIssues.issues()) {
+      String severity = issue.severity();
+      int issueSeverityAsInt = Severity.ALL.indexOf(String.valueOf(severity).trim().toUpperCase());
+      if (issueSeverityAsInt >= failForIssueSeverityAsInt) {
+        LOG.info("Recording issue {} that has a severity of '{}'",
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Info could be pretty spammy here for basic analysis logs.  Probably better to leave this type of information for the preview report.

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+            issue.key(), severity);
+        issueCountToFailFor++;
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Doesn't the preview report contain the issue counts by severity?  I know the number is for the log message but is it important enough to keep around and not simply break out of the loop on the first issue with a high enough severity?

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+      }
+    }
+
+    if (issueCountToFailFor > 0) {
+      String msg = "Project " + project.getName() + " has " + issueCountToFailFor
+          + " issues that are of severity equal or higher than " + failForIssuesWithSeverity;
+      LOG.debug(msg);
+      // only mark failure and fail on PostJobsPhaseHandler.onPostJobsPhase() to ensure other plugins can finish their work, most notably the stash issue reporter plugin
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Long line

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+      markFailure(msg);
+      
+    } else {
+      LOG.info("Project " + project.getName() + " has no issues with severity equal or higher than " + failForIssuesWithSeverity);
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Long line

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java [hunk 18]
```diff
@@ -83,8 +93,11 @@
...
+
   @Override
   public List getExtensions() {
-    return Arrays.asList(ForbiddenConfigurationBreaker.class, QualityGateBreaker.class);
+    return Arrays.asList(ForbiddenConfigurationBreaker.class, QualityGateBreaker.class, BasicIssuesBuildBreaker.class);
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: I'd rather this retain alphabetical order

**Comment by mjdetullio** [comment]
Also, unit tests?

File: src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java (hunk 10 above)
Comment (CSchulz) [inline src/main/java/org/sonar/plugins/buildbreaker/BasicIssuesBuildBreaker.java]: Spaces in second condition around operator

**Comment by CSchulz** [comment]
Documentation update is missing too.

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java (hunk 13 above)
Comment (dspaven) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: Would it not make more sense just to honour sonar.buildbreaker.skip?

**Comment by mjdetullio** [comment]
I merged this myself, but made many changes on top of it.

**Comment by ghenzler** [comment]
@mjdetullio Thanks for taking care of this! (I didn't find the time when you asked for the changes)

**Pull request merged**